### PR TITLE
Iox #25 add integration test pubsub port user

### PR DESCRIPTION
**Pull request by marthtz**
Integration test for subscriber port and publisher port. Single and multi producer.

### CODE REVIEW

File: iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp [hunk 1]
```diff
@@ -42,7 +42,7 @@ static constexpr uint32_t NUM_CHUNKS_IN_POOL = 3 * iox::MAX_RECEIVER_QUEUE_CAPAC
 static constexpr uint32_t SMALL_CHUNK = 128;
 static constexpr uint32_t CHUNK_META_INFO_SIZE = 256;
 static constexpr size_t MEMORY_SIZE = NUM_CHUNKS_IN_POOL * (SMALL_CHUNK + CHUNK_META_INFO_SIZE);
-alignas(64) uint8_t g_memory[MEMORY_SIZE];
+alignas(64) static uint8_t g_memory[MEMORY_SIZE];
```
Comment (bishibashiB) [inline iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp:45]: As an interested reader I am wondering on the origins of the `alignas(64)` (in special this 'magic number' 64).
Somebody shading a light onto the unilluminated, maybe @MatthiasKillat

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_pusher.inl [hunk 2]
```diff
@@ -65,6 +65,9 @@ inline cxx::expected<ChunkQueueError> ChunkQueuePusher<ChunkQueueDataType>::push
                 iox::relative_ptr<mepoo::ChunkManagement>(chunkTupleOut.m_chunkOffset, chunkTupleOut.m_segmentId);
             // this will release the chunk
             auto returnedChunk = mepoo::SharedChunk(chunkManagement);
+            /// we have to set this to true to inform the higher levels that there
+            /// was a chunk lost
+            getMembers()->m_queueHasOverflown.store(true, std::memory_order_relaxed);
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_pusher.inl:70]: Nice one!

File: iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    }
+
+    void SetUp()
+    {
+        for (uint32_t i = 0; i < NUMBER_OF_PUBLISHERS; i++)
+        {
+            std::stringstream publisherAppName;
+            publisherAppName << TEST_PUBLISHER_APP_NAME << i;
+
+            iox::cxx::string<100> processName(TruncateToCapacity, publisherAppName.str().c_str());
```
Comment (mossmaurice) [inline iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp:77]: #253 will introduce the "+ operator" for `cxx::string`;-)

File: iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+    void subscriberThread(uint32_t numberOfPublishers,
+                          SubscriberPortProducerType& subscriberPortRouDi,
+                          SubscriberPortUser& subscriberPortUser)
+    {
+        bool finished{false};
+        optional<CaproMessage> maybeCaproMessage;
+        CaproMessage caproMessage;
+
+        // Wait for publisher to be ready
+        caproMessage = waitForCaproMessage(m_concurrentCaproMessageVector, CaproMessageType::OFFER);
```
Comment (mossmaurice) [inline iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp:173]: `auto caproMessage`?

File: iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+
+        // Subscribe to publisher
+        subscriberPortUser.subscribe();
+        maybeCaproMessage = subscriberPortRouDi.getCaProMessage();
```
Comment (mossmaurice) [inline iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp:177]: `auto maybeCaproMessage`? Then you save line 169 ;)

File: iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+                         PublisherPortUser& publisherPortUser)
+    {
+        optional<CaproMessage> maybeCaproMessage;
+        CaproMessage caproMessage;
+
+        // Publisher offers its service
+        publisherPortUser.offer();
+
+        // Let RouDi change state and send OFFER to subscriber
+        maybeCaproMessage = publisherPortRouDi.getCaProMessage();
```
Comment (mossmaurice) [inline iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp:229]: `auto maybeCaproMessage`?

File: iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+        // Signal the subscriber thread we're done
+        m_publisherRunFinished++;
+    }
+};
+
+TEST_F(PortUser_IntegrationTest, SingleProducer)
+{
+    std::thread subscribingThread(std::bind(&PortUser_IntegrationTest::subscriberThread<SubscriberPortSingleProducer>,
+                                            this,
+                                            1,
```
Comment (mossmaurice) [inline iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp:301]: Magic numbers are magic 🦄 What does this number here mean?

File: iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+        // Let RouDi change state and send OFFER to subscriber
+        maybeCaproMessage = publisherPortRouDi.getCaProMessage();
+
+        if (publisherThreadIndex == 0)
+        {
+            // First publisher thread will sync with subscriber
+            if (maybeCaproMessage.has_value())
+            {
+                caproMessage = maybeCaproMessage.value();
+                auto guardedVector = m_concurrentCaproMessageVector.GetScopeGuard();
```
Comment (budrus) [inline iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp:237]: why do you need the Scope guard here but not some lines below where you send the ACK back?

File: iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+        bool finished{false};
+        CaproMessage caproMessage;
+
+        // Wait until the expected CaPro message has arrived in the shared message vector between subscriber and the
+        // first publisher thread
+        do
+        {
+            // Add delay to allow other thread accessing the shared resource
+            std::this_thread::sleep_for(std::chrono::microseconds(10));
+            {
```
Comment (budrus) [inline iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp:145]: This is a hard pattern. Maybe two of our cuncurrent FiFos would be more straighforward to exchange the CaPro messages between subscriber and the first publisher. But so far I cannot find a problem with this approach

File: iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp (hunk 4 above)
Comment (marthtz) [reply iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp]: Sure. Done.

File: iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp (hunk 5 above)
Comment (marthtz) [reply iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp]: Done.

File: iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp (hunk 6 above)
Comment (marthtz) [reply iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp]: Done.

File: iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp (hunk 7 above)
Comment (marthtz) [reply iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp]: Done.

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_pusher.inl (hunk 2 above)
[the remark this answers is not in the document]
Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_pusher.inl:70]: It's a 1:1 copy as the corresponding commit includes other changes so I couldn't cherry-pick.

File: iceoryx_posh/include/iceoryx_posh/internal/popo/ports/subscriber_port_user.hpp [hunk 10]
```diff
@@ -81,8 +81,8 @@ class SubscriberPortUser
     bool hasLostChunks() noexcept;
 
     /// @todo we first need the new condition variable
-    void attachConditionVariable() noexcept;
```
[the remark this answers is not in the document]
Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/ports/subscriber_port_user.hpp:84]: Of course! Added proper comments.

File: iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+    std::atomic<uint64_t> m_publisherRunFinished{0};
+
+    // Memory objects
+    Allocator m_memoryAllocator{g_memory, MEMORY_SIZE};
+    MePooConfig m_mempoolConfig;
+    MemoryManager m_memoryManager;
+
+    using ConcurrentCaproMessageVector_t = iox::concurrent::smart_lock<vector<CaproMessage, 1>>;
+    ConcurrentCaproMessageVector_t m_concurrentCaproMessageVector;
+    iox::concurrent::smart_lock<vector<CaproMessage, 1>> m_caproMessageRx;
```
[the remark this answers is not in the document]
Comment (marthtz) [reply iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp]: Ah, yeah. Also improved their names.

File: iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+        // Publisher offers its service
+        publisherPortUser.offer();
+
+        // Let RouDi change state and send OFFER to subscriber
+        maybeCaproMessage = publisherPortRouDi.getCaProMessage();
+
+        if (publisherThreadIndex == 0)
+        {
+            // First publisher thread will sync with subscriber
+            if (maybeCaproMessage.has_value())
```
[the remark this answers is not in the document]
Comment (marthtz) [reply iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp:236]: Yes. Catching all maybeCapros now.

File: iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp (hunk 8 above)
Comment (marthtz) [reply iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp]: Must be a leftover. Remove the `GetScopeGuard()` as only single access is needed here so a `->` is sufficient.

**Review by mossmaurice — APPROVED** [review]

File: iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp (hunk 9 above)
Comment (marthtz) [reply iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp:145]: I couldn't make a CaPro message type atomic, hence this approach. I didn't want to mess around with mutex and lock so I used `smart_lock` from iceoryx. Simple handshake between threads. One thread waiting for a specific message before continuing, other thread writing the message. Then vice versa.

Comment (budrus) [reply iceoryx_posh/test/integrationtests/test_popo_port_user_building_blocks.cpp:145]: OK, also not a big issue for the test

**Review by budrus — APPROVED** [review]

File: iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp (hunk 1 above)
Comment (bishibashiB) [reply iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp:45]: maybe something connected to _SC_PAGESIZE (but hard-coded)?

**Pull request merged**